rescue/src/rescue.rs

```diff
@@ -0,0 +1,315 @@
+use crate::inverse_sbox::InverseSboxLayer;
+use crate::util::{binomial, shake256_hash};
+
+use itertools::Itertools;
+use num::{BigUint, One};
+use p3_field::{PrimeField, PrimeField64};
+use p3_symmetric::permutation::{ArrayPermutation, CryptographicPermutation, MDSPermutation};
+use p3_util::ceil_div_usize;
+use rand::distributions::Standard;
+use rand::prelude::Distribution;
+use rand::Rng;
+use std::marker::PhantomData;
+
+#[derive(Clone)]
+pub struct Rescue<
+    F,
+    MDS,
+    ISL,
+    const WIDTH: usize,
+    const CAPACITY: usize,
+    const ALPHA: u64,
+    const SEC_LEVEL: usize,
+> where
+    F: PrimeField,
+    MDS: MDSPermutation<F, WIDTH>,
+    ISL: InverseSboxLayer<F, WIDTH, ALPHA>,
+{
+    num_rounds: usize,
+    mds: MDS,
+    isl: ISL,
+    round_constants: Vec<F>,
+
+    _phantom_f: PhantomData<F>,
+}
+
+impl<
+        F,
+        MDS,
+        ISL,
+        const WIDTH: usize,
+        const CAPACITY: usize,
```

Comment:
Similar here, looks like it's used in `get_round_constants_rescue_prime`, but otherwise I don't think we need it since we can implement `ArrayPermutation` and leave sponge logic to `PaddingFreeSponge`. So I think it could be an argument to certain constructors, but in general shouldn't need to be specified (say if the user wants to manually supply the constants etc).

Reply:
Oh good idea! Much cleaner with fewer generics 😅